File: file_detect copy.c
#include <uapi/linux/ptrace.h>
#include <linux/sched.h>

/*
Event payload we define to send to user space.
- This is NOT a kernel-provided struct; it's our custom payload format.
- It will be delivered to user space as-is via a perf(per cpu) ring buffer (map).
 */
struct file_event_t {
    u32 pid;                  // using (bpf_get_current_pid_tgid() >> 32) yields TGID (process ID).
    u32 uid;                  // Caller’s UID. We use the lower 32 bits of bpf_get_current_uid_gid().
    char comm[TASK_COMM_LEN]; // Current task's command name. Up to 16 bytes (including NUL).
    char fname[256];          // File path passed to openat (read from user memory).
};

/*
BCC macro-BPF_PERF_OUTPUT: define a perf ring buffer map.
- Name: file_events
- The eBPF program pushes events into this buffer via perf_submit(),
- and user space (Python BCC) consumes them with open_perf_buffer().
*/
BPF_PERF_OUTPUT(file_events);

/*
TRACEPOINT_PROBE(syscalls, sys_enter_openat)
- eBPF function attached to the sys_enter_openat tracepoint.
- 'args' here is a pointer to an auto-generated struct provided by BCC,
exposing the tracepoint fields (dfd, filename, flags, mode, etc.) as 'args->field_name
!!! The exact shape of this struct is derived from the kernel's tracepoint !!! 
    to find format metadata:
    sudo cat /sys/kernel/debug/tracing/events/syscalls/sys_enter_openat/format
    or
    sudo bpftrace -lv 'tracepoint:syscalls:sys_enter_openat'
*/
TRACEPOINT_PROBE(syscalls, sys_enter_openat) {
    // Allocate one event record on the eBPF stack (~280B, below the 512B stack limit).
    struct file_event_t evt = {};

    /*
    Fetch pid/tgid, uid/gid, and comm using eBPF helpers.
    bpf_get_current_pid_tgid():
    - returns 64 bits: upper 32 = TGID (thread group ID == process ID), lower 32 = PID (thread ID)
    - Here we store TGID into evt.pid (>> 32).
    If you need TID (thread ID), cast to (u32) to use the lower 32 bits.
    */
    evt.pid = bpf_get_current_pid_tgid() >> 32;

    /*
    bpf_get_current_uid_gid():
    - returns 64 bits: lower 32 = UID, upper 32 = GID
    - evt.uid is u32, so only the lower 32 bits (UID) are stored.
    */
    evt.uid = bpf_get_current_uid_gid();

    /*
    Copy current task's comm (process name) into evt.comm.
    - The helper guarantees NUL termination.
    */
    bpf_get_current_comm(&evt.comm, sizeof(evt.comm));

    /*
    Safely read the user pointer 'filename' passed to openat.
    - bpf_probe_read_user_str(): reads a NUL-terminated string from user memory.
    - Copies up to the given size and guarantees NUL termination.
    - Returns a negative error code on failure; here we just attempt the read.
    'args->filename' is a field from the *tracepoint struct*.
    (With kprobes, you would fetch arguments from registers instead.)
    (we can find the functions that can attach kprobes with command below.)
    sudo cat /sys/kernel/debug/tracing/available_filter_functions | less
    */
    bpf_probe_read_user_str(&evt.fname, sizeof(evt.fname), args->filename);

    /*
    Submit the collected event to the perf ring buffer.
    - Passing 'args' as the first argument is a common BCC convention for ctx(context pointer).
    - User space consumes it via b["file_events"].open_perf_buffer(...).
    */
    file_events.perf_submit(args, &evt, sizeof(evt));
    return 0;
}

File: file_detect copy.py
# BCC (BPF Compiler Collection) Python bindings
from bcc import BPF

# --------------------------------------------------------------------
# 1) Load eBPF C source
#    - Read file_detect.c from the same directory.
#    - file_detect.c defines our struct file_event_t and
#      TRACEPOINT_PROBE(syscalls, sys_enter_openat).
#      - The kernel tracepoint provides the filename argument, and the
#        eBPF program sends events to a perf ring buffer.
# --------------------------------------------------------------------
with open("file_detect.c") as f:
    bpf_source = f.read()

# --------------------------------------------------------------------
# 2) Compile & load eBPF
#    - BCC uses LLVM/Clang to compile the C code and then loads the
#      resulting BPF program into the kernel.
# --------------------------------------------------------------------
b = BPF(text=bpf_source)

# --------------------------------------------------------------------
# 3) Attach the tracepoint
#    - For TRACEPOINT_PROBE(syscalls, sys_enter_openat) in C, the
#      function name in Python is "tracepoint__syscalls__sys_enter_openat".
#    - Some environments autoload, but here we attach explicitly.
# --------------------------------------------------------------------
# try:
#     b.attach_tracepoint(tp="syscalls:sys_enter_openat",
#                         fn_name="tracepoint__syscalls__sys_enter_openat")
# except Exception as e:
#     print(f"[!] Failed to attach tracepoint: {e}")
#     raise SystemExit(1)

# --------------------------------------------------------------------
# 4) Event callback
#    - The eBPF side defined BPF_PERF_OUTPUT(file_events), and pushes
#      struct file_event_t via file_events.perf_submit().
#    - b["file_events"].event(data) parses that C struct layout directly,
#      so we can access event.pid / event.uid / event.comm / event.fname.
# --------------------------------------------------------------------
def on_file_event(cpu, data, size):
    event = b["file_events"].event(data)

    # Safely convert C char arrays to Python strings
    comm  = event.comm.decode(errors="ignore")
    fname = event.fname.decode(errors="ignore").rstrip("\x00")

    # Watch target: detect access to /etc/shadow
    if fname == "/etc/shadow":
        print(f"[*] ALERT: /etc/shadow accessed! "
              f"(PID: {event.pid}, Proc: {comm}, UID: {event.uid})")
    else:
        # Informational log (remove if too noisy)
        print(f"[i] Skipped: {fname} "
              f"(PID: {event.pid}, Proc: {comm}, UID: {event.uid})")

# --------------------------------------------------------------------
# 5) Open perf buffer & polling loop
#    - Register the callback with open_perf_buffer().
#    - perf_buffer_poll() blocks and dispatches events to the callback.
#    - Keep looping so the program stays alive and processes events.
# --------------------------------------------------------------------
b["file_events"].open_perf_buffer(on_file_event)

print("[*] Monitoring for /etc/shadow access... (Press Ctrl+C to stop)")
try:
    while True:
        b.perf_buffer_poll()
except KeyboardInterrupt:
    print("\n[!] Detaching BPF program and exiting...")
finally:
    # Deleting 'b' lets BCC detach the tracepoint and clean up maps
    del b
    print("[*] BPF program detached successfully.")
